File: src/test/java/de/pflugradts/passbird/application/boot/setup/PassbirdSetupTest.java
package de.pflugradts.passbird.application.boot.setup;

import de.pflugradts.passbird.application.KeyStoreAdapterPort;
import de.pflugradts.passbird.application.UserInterfaceAdapterPort;
import de.pflugradts.passbird.application.UserInterfaceAdapterPortFaker;
import de.pflugradts.passbird.application.configuration.Configuration;
import de.pflugradts.passbird.application.configuration.ConfigurationSync;
import de.pflugradts.passbird.application.configuration.MockitoConfigurationFaker;
import de.pflugradts.passbird.application.failurehandling.FailureCollector;
import de.pflugradts.passbird.application.util.FileFaker;
import de.pflugradts.passbird.application.util.PathFaker;
import de.pflugradts.passbird.application.util.SystemOperation;
import de.pflugradts.passbird.application.util.SystemOperationFaker;
import de.pflugradts.passbird.domain.model.transfer.Input;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.Spy;
import org.mockito.junit.jupiter.MockitoExtension;

import java.nio.file.Path;

import static de.pflugradts.passbird.domain.model.transfer.InputFakerKt.fakeInput;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;
import static org.mockito.BDDMockito.willDoNothing;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.mock;

@ExtendWith(MockitoExtension.class)
class PassbirdSetupTest {

    @Spy
    private SetupGuide setupGuide = new SetupGuide(mock(UserInterfaceAdapterPort.class));
    @Mock
    private ConfigurationSync configurationSync;
    @Mock
    private FailureCollector failureCollector;
    @Mock
    private Configuration configuration;
    @Mock
    private KeyStoreAdapterPort keyStoreAdapterPort;
    @Mock
    private UserInterfaceAdapterPort userInterfaceAdapterPort;
    @Mock
    private SystemOperation systemOperation;
    @InjectMocks
    private PassbirdSetup passbirdSetup;

    @Captor
    private ArgumentCaptor<Path> captor;

    @Test
    void shouldRunConfigTemplateRoute() {
        // given
        final var configurationDirectory = "tmp";
        final var password = fakeInput("p4s5w0rD");
        MockitoConfigurationFaker.faker()
                .forInstance(configuration)
                .withConfigurationTemplate()
                .withPasswordStoreLocation(configurationDirectory)
                .withKeyStoreLocation(configurationDirectory)
                .fake();
        UserInterfaceAdapterPortFaker.faker()
                .forInstance(userInterfaceAdapterPort)
                .withTheseSecureInputs(password, password)
                .withReceiveConfirmation(true).fake();
        willDoNothing().given(configurationSync).sync(configurationDirectory);
        givenValidDirectory(configurationDirectory);

        // when
        passbirdSetup.boot();

        // then
        then(setupGuide).should().sendWelcome();
        then(setupGuide).should().sendConfigTemplateRouteInformation();
        then(setupGuide).should().sendInputPath("configuration");
        then(setupGuide).should().sendCreateKeyStoreInformation();
        then(setupGuide).should().sendRestart();
        then(setupGuide).should().sendGoodbye();
        then(systemOperation).should().exit();
    }

    @Test
    void shouldAbortUnconfirmedConfigTemplateRoute() {
        // given
        MockitoConfigurationFaker.faker()
                .forInstance(configuration)
                .withConfigurationTemplate()
                .fake();
        UserInterfaceAdapterPortFaker.faker()
                .forInstance(userInterfaceAdapterPort)
                .withReceiveConfirmation(false).fake();

        // when
        passbirdSetup.boot();

        // then
        then(setupGuide).should().sendWelcome();
        then(setupGuide).should().sendConfigTemplateRouteInformation();
        then(setupGuide).should().sendGoodbye();
        then(systemOperation).should().exit();
    }

    @Test
    void shouldRunConfigKeyStoreRoute() {
        // given
        final var configurationDirectory = "tmp";
        final var password = fakeInput("p4s5w0rD");
        MockitoConfigurationFaker.faker()
                .forInstance(configuration)
                .withKeyStoreLocation(configurationDirectory)
                .fake();
        UserInterfaceAdapterPortFaker.faker()
                .forInstance(userInterfaceAdapterPort)
                .withTheseSecureInputs(password, password)
                .withReceiveConfirmation(true).fake();
        givenValidDirectory(configurationDirectory);

        // when
        passbirdSetup.boot();

        // then
        then(setupGuide).should().sendWelcome();
        then(setupGuide).should().sendConfigKeyStoreRouteInformation(configurationDirectory);
        then(setupGuide).should().sendInputPath("keystore");
        then(setupGuide).should().sendCreateKeyStoreInformation();
        then(setupGuide).should().sendRestart();
        then(setupGuide).should().sendGoodbye();
        then(systemOperation).should().exit();
    }

    @Test
    void shouldAbortUnconfirmedConfigKeyStoreRoute() {
        // given
        final var configurationDirectory = "tmp";
        MockitoConfigurationFaker.faker()
                .forInstance(configuration)
                .withKeyStoreLocation(configurationDirectory)
                .fake();
        UserInterfaceAdapterPortFaker.faker()
                .forInstance(userInterfaceAdapterPort)
                .withReceiveConfirmation(false).fake();

        // when
        passbirdSetup.boot();

        // then
        then(setupGuide).should().sendWelcome();
        then(setupGuide).should().sendConfigKeyStoreRouteInformation(configurationDirectory);
        then(setupGuide).should().sendGoodbye();
        then(systemOperation).should().exit();
    }

    @Test
    void shouldAcceptCorrectedDirectory() {
        // given
        MockitoAnnotations.initMocks(this);
        final var invalidConfigurationDirectory = "/dev/null";
        final var validDirectory = fakeInput("tmp");
        final var password = fakeInput("p4s5w0rD");
        MockitoConfigurationFaker.faker()
                .forInstance(configuration)
                .withKeyStoreLocation(invalidConfigurationDirectory)
                .fake();
        UserInterfaceAdapterPortFaker.faker()
                .forInstance(userInterfaceAdapterPort)
                .withTheseInputs(validDirectory)
                .withTheseSecureInputs(password, password)
                .withReceiveConfirmation(true).fake();
        doNothing().when(keyStoreAdapterPort).storeKey(any(), captor.capture());
        givenInvalidDirectory(invalidConfigurationDirectory);
        givenValidDirectory(validDirectory.getBytes().asString());

        // when
        passbirdSetup.boot();

        // then
        then(setupGuide).should().sendCreateKeyStoreInformation();
        assertThat(captor.getValue().toString()).startsWith(validDirectory.getBytes().asString());
        then(setupGuide).should().sendRestart();
        then(setupGuide).should().sendGoodbye();
        then(systemOperation).should().exit();
    }

    @Test
    void shouldCreateKeystoreWithMatchingPasswordInput() {
        // given
        final var configurationDirectory = "tmp";
        final var passwordMismatch1 = fakeInput("bassword");
        final var passwordMismatch2 = fakeInput("guessword");
        final var emptyPassword =  Input.Companion.emptyInput();
        final var passwordMatched = fakeInput("p4s5w0rD");
        MockitoConfigurationFaker.faker()
                .forInstance(configuration)
                .withKeyStoreLocation(configurationDirectory)
                .fake();
        UserInterfaceAdapterPortFaker.faker()
                .forInstance(userInterfaceAdapterPort)
                .withTheseSecureInputs(
                        passwordMismatch1,
                        passwordMismatch2,
                        emptyPassword,
                        emptyPassword,
                        passwordMatched,
                        passwordMatched)
                .withReceiveConfirmation(true).fake();
        givenValidDirectory(configurationDirectory);

        // when
        passbirdSetup.boot();

        // then
        then(setupGuide).should().sendCreateKeyStoreInformation();
        then(keyStoreAdapterPort).should().storeKey(eq(passwordMatched.getBytes().toChars()), any());
        then(setupGuide).should().sendRestart();
        then(setupGuide).should().sendGoodbye();
        then(systemOperation).should().exit();
    }

    private void givenValidDirectory(final String directory) {
        given(systemOperation.resolvePath(eq(directory), anyString())).willCallRealMethod();
        final var parent = FileFaker.faker()
                .fakeFile()
                .withDirectoryProperty(true)
                .withExistsProperty(true).fake();
        final var file = FileFaker.faker()
                .fakeFile()
                .withDirectoryProperty(true)
                .withParentFile(parent).fake();
        final var path = PathFaker.faker()
                .fakePath()
                .withFileRepresentation(file).fake();
        SystemOperationFaker.faker()
                .forInstance(systemOperation)
                .withPath(directory, path).fake();
    }

    private void givenInvalidDirectory(final String directory) {
        final var file = FileFaker.faker()
                .fakeFile()
                .withDirectoryProperty(false).fake();
        final var path = PathFaker.faker()
                .fakePath()
                .withFileRepresentation(file).fake();
        SystemOperationFaker.faker()
                .forInstance(systemOperation)
                .withPath(directory, path).fake();
    }

}

File: src/test/kotlin/de/pflugradts/passbird/application/boot/setup/PassbirdSetupTest.kt
package de.pflugradts.passbird.application.boot.setup

import de.pflugradts.kotlinextensions.TryResult.Companion.failure
import de.pflugradts.kotlinextensions.TryResult.Companion.success
import de.pflugradts.passbird.application.KeyStoreAdapterPort
import de.pflugradts.passbird.application.UserInterfaceAdapterPort
import de.pflugradts.passbird.application.configuration.Configuration
import de.pflugradts.passbird.application.configuration.ConfigurationSync
import de.pflugradts.passbird.application.configuration.ReadableConfiguration
import de.pflugradts.passbird.application.configuration.fakeConfiguration
import de.pflugradts.passbird.application.fakeUserInterfaceAdapterPort
import de.pflugradts.passbird.application.toDirectory
import de.pflugradts.passbird.application.util.SystemOperation
import de.pflugradts.passbird.application.util.fakePath
import de.pflugradts.passbird.application.util.fakeSystemOperation
import de.pflugradts.passbird.domain.model.shell.Shell.Companion.shellOf
import de.pflugradts.passbird.domain.model.transfer.Input.Companion.emptyInput
import de.pflugradts.passbird.domain.model.transfer.Output.Companion.outputOf
import de.pflugradts.passbird.domain.model.transfer.fakeInput
import io.mockk.Called
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.spyk
import io.mockk.verify
import io.mockk.verifyOrder
import org.junit.jupiter.api.Test
import strikt.api.expectThat
import strikt.assertions.isEqualTo
import strikt.assertions.isNotEqualTo
import java.nio.file.Path
import kotlin.io.path.name

private const val VALID_DIRECTORY = "tmp"

class PassbirdSetupTest {
    private val userInterfaceAdapterPort = mockk<UserInterfaceAdapterPort>()
    private val setupGuide = spyk(SetupGuide(userInterfaceAdapterPort))
    private val configurationSync = mockk<ConfigurationSync>()
    private val configuration = mockk<Configuration>(relaxed = true)
    private val keyStoreAdapterPort = mockk<KeyStoreAdapterPort>()
    private val systemOperation = mockk<SystemOperation>()
    private val passbirdSetup = PassbirdSetup(
        setupGuide = setupGuide,
        configurationSync = configurationSync,
        configuration = configuration,
        keyStoreAdapterPort = keyStoreAdapterPort,
        userInterfaceAdapterPort = userInterfaceAdapterPort,
        systemOperation = systemOperation,
    )

    @Test
    fun `should run config template route`() {
        // given
        val configurationDirectory = VALID_DIRECTORY
        val password1 = fakeInput("p4s5w0rD")
        val password2 = fakeInput("p4s5w0rD")
        val pathSlot = slot<Path>()
        fakeConfiguration(
            instance = configuration,
            withConfigurationTemplate = true,
            withPasswordTreeLocation = configurationDirectory,
            withKeyStoreLocation = configurationDirectory,
        )
        fakeUserInterfaceAdapterPort(
            instance = userInterfaceAdapterPort,
            withTheseSecureInputs = listOf(password1, password2),
            withReceiveConfirmation = true,
        )
        fakeSystemOperation(
            instance = systemOperation,
            withPaths = listOf(Pair(VALID_DIRECTORY, fakePath(exists = true, isDirectory = true))),
        )
        every { configurationSync.sync(configurationDirectory.toDirectory()) } returns success(Unit)
        every { keyStoreAdapterPort.storeKey(eq(password1.shell.toPlainShell()), capture(pathSlot)) } returns Unit

        // when
        passbirdSetup.boot()

        // then
        verify(exactly = 1) { setupGuide.sendWelcome() }
        verify(exactly = 1) { setupGuide.sendConfigTemplateRouteInformation() }
        verify(exactly = 1) { setupGuide.sendInputPath("configuration") }
        verify(exactly = 1) { setupGuide.sendCreateKeyStoreInformation() }
        verify(exactly = 0) { setupGuide.sendNonMatchingInputs() }
        expectThat(pathSlot.captured.fileName.name) isEqualTo ReadableConfiguration.KEYSTORE_FILENAME
        expectThat(pathSlot.captured.parent.name) isEqualTo configurationDirectory
        expectThat(password1.shell.asString()) isNotEqualTo "p4s5w0rD"
        expectThat(password2.shell.asString()) isNotEqualTo "p4s5w0rD"
        verify(exactly = 1) { setupGuide.sendRestart() }
        verify(exactly = 1) { setupGuide.sendGoodbye() }
        verify(exactly = 1) { systemOperation.exit() }
    }

    @Test
    fun `should abort unconfirmed config template route`() {
        // given
        fakeConfiguration(instance = configuration, withConfigurationTemplate = true)
        fakeUserInterfaceAdapterPort(instance = userInterfaceAdapterPort, withReceiveConfirmation = false)
        fakeSystemOperation(instance = systemOperation)

        // when
        passbirdSetup.boot()

        // then
        verify(exactly = 1) { setupGuide.sendWelcome() }
        verify(exactly = 1) { setupGuide.sendConfigTemplateRouteInformation() }
        verify(exactly = 1) { setupGuide.sendGoodbye() }
        verify(exactly = 1) { systemOperation.exit() }
        verify { configurationSync wasNot Called }
        verify { keyStoreAdapterPort wasNot Called }
    }

    @Test
    fun `should abort config template route after failed configuration sync`() {
        // given
        val configurationDirectory = VALID_DIRECTORY
        fakeConfiguration(
            instance = configuration,
            withConfigurationTemplate = true,
            withPasswordTreeLocation = configurationDirectory,
        )
        fakeUserInterfaceAdapterPort(instance = userInterfaceAdapterPort, withReceiveConfirmation = true)
        fakeSystemOperation(
            instance = systemOperation,
            withPaths = listOf(Pair(VALID_DIRECTORY, fakePath(exists = true, isDirectory = true))),
        )
        every {
            configurationSync.sync(configurationDirectory.toDirectory())
        } returns failure(IllegalStateException("disk full"))

        // when
        passbirdSetup.boot()

        // then
        verify(exactly = 1) { setupGuide.sendWelcome() }
        verify(exactly = 1) { setupGuide.sendConfigTemplateRouteInformation() }
        verify(exactly = 1) { setupGuide.sendInputPath("configuration") }
        verify(exactly = 0) { setupGuide.sendCreateKeyStoreInformation() }
        verify(exactly = 0) { setupGuide.sendRestart() }
        verify(exactly = 1) { setupGuide.sendGoodbye() }
        verify(exactly = 1) { systemOperation.exit() }
        verify { keyStoreAdapterPort wasNot Called }
    }

    @Test
    fun `should run config key store route`() {
        // given
        val configurationDirectory = VALID_DIRECTORY
        val password1 = fakeInput("p4s5w0rD")
        val password2 = fakeInput("p4s5w0rD")
        val pathSlot = slot<Path>()
        fakeConfiguration(instance = configuration, withKeyStoreLocation = configurationDirectory)
        fakeUserInterfaceAdapterPort(
            instance = userInterfaceAdapterPort,
            withTheseSecureInputs = listOf(password1, password2),
            withReceiveConfirmation = true,
        )
        fakeSystemOperation(
            instance = systemOperation,
            withPaths = listOf(Pair(VALID_DIRECTORY, fakePath(exists = true, isDirectory = true))),
        )
        every { keyStoreAdapterPort.storeKey(eq(password1.shell.toPlainShell()), capture(pathSlot)) } returns Unit

        // when
        passbirdSetup.boot()

        // then
        verify(exactly = 1) { setupGuide.sendWelcome() }
        verify(exactly = 1) { setupGuide.sendConfigKeyStoreRouteInformation(configurationDirectory) }
        verify(exactly = 1) { setupGuide.sendInputPath("keystore") }
        verify(exactly = 1) { setupGuide.sendCreateKeyStoreInformation() }
        verify(exactly = 0) { setupGuide.sendNonMatchingInputs() }
        expectThat(pathSlot.captured.fileName.name) isEqualTo ReadableConfiguration.KEYSTORE_FILENAME
        expectThat(pathSlot.captured.parent.name) isEqualTo configurationDirectory
        expectThat(password1.shell.asString()) isNotEqualTo "p4s5w0rD"
        expectThat(password2.shell.asString()) isNotEqualTo "p4s5w0rD"
        verify(exactly = 1) { setupGuide.sendRestart() }
        verify(exactly = 1) { setupGuide.sendGoodbye() }
        verify(exactly = 1) { systemOperation.exit() }
        verify { configurationSync wasNot Called }
    }

    @Test
    fun `should abort unconfirmed config key store route`() {
        // given
        val configurationDirectory = VALID_DIRECTORY
        fakeConfiguration(instance = configuration, withKeyStoreLocation = configurationDirectory)
        fakeUserInterfaceAdapterPort(instance = userInterfaceAdapterPort, withReceiveConfirmation = false)
        fakeSystemOperation(instance = systemOperation)

        // when
        passbirdSetup.boot()

        // then
        verify(exactly = 1) { setupGuide.sendWelcome() }
        verify(exactly = 1) { setupGuide.sendConfigKeyStoreRouteInformation(configurationDirectory) }
        verify(exactly = 1) { setupGuide.sendGoodbye() }
        verify(exactly = 1) { systemOperation.exit() }
        verify { configurationSync wasNot Called }
        verify { keyStoreAdapterPort wasNot Called }
    }

    @Test
    fun `should accept corrected directory`() {
        // given
        val invalidConfigurationDirectory = "/dev/null"
        val nonexistentConfigurationDirectory = "/dev/none"
        val validDirectory = fakeInput(VALID_DIRECTORY)
        val password1 = fakeInput("p4s5w0rD")
        val password2 = fakeInput("p4s5w0rD")
        val pathSlot = slot<Path>()
        fakeConfiguration(instance = configuration, withKeyStoreLocation = invalidConfigurationDirectory)
        fakeUserInterfaceAdapterPort(
            instance = userInterfaceAdapterPort,
            withTheseInputs = listOf(fakeInput(nonexistentConfigurationDirectory), validDirectory),
            withTheseSecureInputs = listOf(password1, password2),
            withReceiveConfirmation = true,
        )
        fakeSystemOperation(
            instance = systemOperation,
            withPaths = listOf(
                Pair(VALID_DIRECTORY, fakePath(exists = true, isDirectory = true)),
                Pair(invalidConfigurationDirectory, fakePath(exists = true, isDirectory = false)),
                Pair(nonexistentConfigurationDirectory, fakePath(exists = false, isDirectory = true)),
            ),
        )
        every { keyStoreAdapterPort.storeKey(eq(password1.shell.toPlainShell()), capture(pathSlot)) } returns Unit

        // when
        passbirdSetup.boot()

        // then
        verify(exactly = 1) { setupGuide.sendWelcome() }
        expectThat(pathSlot.captured.fileName.name) isEqualTo ReadableConfiguration.KEYSTORE_FILENAME
        expectThat(pathSlot.captured.parent.name) isEqualTo VALID_DIRECTORY
        expectThat(password1.shell.asString()) isNotEqualTo "p4s5w0rD"
        expectThat(password2.shell.asString()) isNotEqualTo "p4s5w0rD"
        verify(exactly = 1) { setupGuide.sendGoodbye() }
        verify(exactly = 1) { systemOperation.exit() }
    }

    @Test
    fun `should create key store with matching password input`() {
        // given
        val configurationDirectory = VALID_DIRECTORY
        val passwordMismatch1 = fakeInput("bassword")
        val passwordMismatch2 = fakeInput("guessword")
        val emptyPassword1 = emptyInput()
        val emptyPassword2 = emptyInput()
        val passwordMatched1 = fakeInput("p4s5w0rD")
        val passwordMatched2 = fakeInput("p4s5w0rD")
        val pathSlot = slot<Path>()
        fakeConfiguration(instance = configuration, withKeyStoreLocation = configurationDirectory)
        fakeUserInterfaceAdapterPort(
            instance = userInterfaceAdapterPort,
            withTheseSecureInputs = listOf(
                passwordMismatch1,
                passwordMismatch2,
                emptyPassword1,
                emptyPassword2,
                passwordMatched1,
                passwordMatched2,
            ),
            withReceiveConfirmation = true,
        )
        fakeSystemOperation(
            instance = systemOperation,
            withPaths = listOf(Pair(VALID_DIRECTORY, fakePath(exists = true, isDirectory = true))),
        )
        every { keyStoreAdapterPort.storeKey(eq(passwordMatched1.shell.toPlainShell()), capture(pathSlot)) } returns Unit

        // when
        passbirdSetup.boot()

        // then
        verify(exactly = 1) { setupGuide.sendWelcome() }
        verify(exactly = 1) { setupGuide.sendConfigKeyStoreRouteInformation(configurationDirectory) }
        verify(exactly = 1) { setupGuide.sendInputPath("keystore") }
        verify(exactly = 1) { setupGuide.sendCreateKeyStoreInformation() }
        verify(exactly = 2) { setupGuide.sendNonMatchingInputs() }
        expectThat(pathSlot.captured.fileName.name) isEqualTo ReadableConfiguration.KEYSTORE_FILENAME
        expectThat(pathSlot.captured.parent.name) isEqualTo configurationDirectory
        expectThat(passwordMismatch1.shell.asString()) isNotEqualTo "bassword"
        expectThat(passwordMismatch2.shell.asString()) isNotEqualTo "guessword"
        expectThat(passwordMatched1.shell.asString()) isNotEqualTo "p4s5w0rD"
        expectThat(passwordMatched2.shell.asString()) isNotEqualTo "p4s5w0rD"
        verify(exactly = 1) { setupGuide.sendRestart() }
        verify(exactly = 1) { setupGuide.sendGoodbye() }
        verify(exactly = 1) { systemOperation.exit() }
        verify { configurationSync wasNot Called }
    }

    @Test
    fun `should only warn about non matching password input after the first failed attempt`() {
        // given
        val configurationDirectory = VALID_DIRECTORY
        val passwordMismatch1 = fakeInput("bassword")
        val passwordMismatch2 = fakeInput("guessword")
        val passwordMatched1 = fakeInput("p4s5w0rD")
        val passwordMatched2 = fakeInput("p4s5w0rD")
        val pathSlot = slot<Path>()
        fakeConfiguration(instance = configuration, withKeyStoreLocation = configurationDirectory)
        fakeUserInterfaceAdapterPort(
            instance = userInterfaceAdapterPort,
            withTheseSecureInputs = listOf(
                passwordMismatch1,
                passwordMismatch2,
                passwordMatched1,
                passwordMatched2,
            ),
            withReceiveConfirmation = true,
        )
        fakeSystemOperation(
            instance = systemOperation,
            withPaths = listOf(Pair(VALID_DIRECTORY, fakePath(exists = true, isDirectory = true))),
        )
        every { keyStoreAdapterPort.storeKey(eq(passwordMatched1.shell.toPlainShell()), capture(pathSlot)) } returns Unit

        // when
        passbirdSetup.boot()

        // then
        verifyOrder {
            userInterfaceAdapterPort.receiveSecurely(outputOf(shellOf("first input: ")))
            userInterfaceAdapterPort.receiveSecurely(outputOf(shellOf("second input: ")))
            userInterfaceAdapterPort.send(outputOf(shellOf("Your inputs do not match, please repeat.")))
            userInterfaceAdapterPort.receiveSecurely(outputOf(shellOf("first input: ")))
            userInterfaceAdapterPort.receiveSecurely(outputOf(shellOf("second input: ")))
        }
        expectThat(pathSlot.captured.fileName.name) isEqualTo ReadableConfiguration.KEYSTORE_FILENAME
        expectThat(pathSlot.captured.parent.name) isEqualTo configurationDirectory
        expectThat(passwordMismatch1.shell.asString()) isNotEqualTo "bassword"
        expectThat(passwordMismatch2.shell.asString()) isNotEqualTo "guessword"
        expectThat(passwordMatched1.shell.asString()) isNotEqualTo "p4s5w0rD"
        expectThat(passwordMatched2.shell.asString()) isNotEqualTo "p4s5w0rD"
    }
}
